sky/serve/controller.py

```diff
@@ -157,6 +158,67 @@ async def update_service(request: fastapi.Request) -> fastapi.Response:
                 return responses.JSONResponse(content={'message': 'Error'},
                                               status_code=500)
 
+        @self._app.post('/controller/terminate_replica')
+        async def terminate_replica(
+                request: fastapi.Request) -> fastapi.Response:
+            request_data = await request.json()
+            replica_id = request_data['replica_id']
+            assert isinstance(replica_id,
+                              int), 'Error: replica ID must be an integer.'
+            purge = request_data['purge']
+            assert isinstance(purge, bool), 'Error: purge must be a boolean.'
+            replica_info = serve_state.get_replica_info_from_id(
+                self._service_name, replica_id)
+            assert replica_info is not None, (f'Error: replica '
+                                              f'{replica_id} does not exist.')
+            replica_status = replica_info.status
+
+            if replica_status == serve_state.ReplicaStatus.SHUTTING_DOWN:
+                return fastapi.Response(
+                    status_code=200,
+                    content={
+                        'message': f'Already scheduled to terminate '
+                                   f'replica {replica_id} of service '
+                                   f'{self._service_name!r}.'
+                    })
+
+            if replica_status in serve_state.ReplicaStatus.failed_statuses():
+                if purge:
+                    self._replica_manager.scale_down(replica_id, purge=True)
```

Comment:
nit: can we combine this branch and the `not replica_status in serve_state.ReplicaStatus.failed_statuses()` branch together?